compiler/optimizer/abstractinterpreter/AbsValue.hpp

```diff
@@ -0,0 +1,160 @@
+ *******************************************************************************/
+
+#ifndef ABS_VALUE_INCL
+#define ABS_VALUE_INCL
+
+#include "optimizer/VPConstraint.hpp"
+#include "il/OMRDataTypes.hpp"
+#include "optimizer/ValuePropagation.hpp"
+
+/**
+ * AbsValue is the basic unit used to perform abstract interpretation.
+ */
+class AbsValue
+   {
+   public:
+   AbsValue(TR::VPConstraint *constraint, TR::DataType dataType, bool isDummy=false);
+   AbsValue(AbsValue* other);
+
+   /**
+    * @brief Merge with another AbsValue. 
+    * Note: This is in-place merge. 
+    * Other should have the exactly SAME dataType as self. 
+    * Any type promotion, such as Int16 => Int32 should be handled outside before calling merge() since this is not handled inside this method.
+    *
+    * @param other AbsValue*
+    * @param vp OMR::ValuePropagation* 
+    * @return AbsValue*
+    * 
+    * The return value is self or NULL.
+    * When succeeding to merge, it returns self.
+    * NULL denotes the merge fails (when merging with different dataTypes or merging with a Dummy AbsValue).
+    */
+   AbsValue* merge(AbsValue *other, OMR::ValuePropagation *vp);
+
+   /**
+    * @brief Check whether the AbsValue has no constraint
+    *
+    * @return bool
+    */
+   bool isTop() { return _constraint == NULL; }
+
+   /**
+    * @brief Set the constriant to unknown.
+    *
+    * @return void
+    */
+   void setToTop() { _constraint = NULL; }
+
+   /**
+    * @brief Check if the AbsValue is only used to take a slot as the second word of the 64-bit data types
+    *
+    * @return bool
+    */
+   bool isDummy() { return _isDummy; }
+
+   /**
+    * @brief Check if the AbsValue is a parameter.
+    *
+    * @return bool
+    */
+   bool isParameter() { return _paramPos >= 0; }
+
+   /**
+    * @brief Check if the AbsValue is 'this' (the implicit parameter.)
+    *
+    * @return bool
+    */   
+   bool isImplicitParameter() { return _paramPos == 0 && _isImplicitParameter; }
+
+   /**
+    * A set of methods for creating different kinds of AbsValue.
+    * OMR::ValuePropagation is required as a parameter since we are using TR::VPConstraint as the constraint.
+    * Though it has nothing to do with the Value Propagation optimization.
+    */
+   static AbsValue* create(TR::VPConstraint *constraint, TR::DataType dataType, TR::Region& region);
+
+   static AbsValue* create(AbsValue* other, TR::Region& region);
+   
+   static AbsValue* createClassObject(TR_OpaqueClassBlock* opaqueClass, bool mustBeNonNull, TR::Region& region, OMR::ValuePropagation* vp);
+
+   static AbsValue* createNullObject(TR::Region& region, OMR::ValuePropagation* vp);
+   static AbsValue* createArrayObject(TR_OpaqueClassBlock* arrayClass, bool mustBeNonNull, int32_t lengthLow, int32_t lengthHigh, int32_t elementSize, TR::Region& region, OMR::ValuePropagation* vp);
+   
+   static AbsValue* createStringConst(TR::SymbolReference* symRef, TR::Region& region, OMR::ValuePropagation* vp);
```

Comment:
I don't know that classes, objects and arrays are that java specific? They map to the same concepts in VP and those are not Java-specific. A language doesn't have to use them, but they are generally useful for a lot of languages no?

Reply:
I happen to think the same concepts in VP _are_ Java-specific. It's true that arrays and objects are concepts applicable to many languages at a high level. However, in my experience, the mechanics of how these are currently represented in several parts of the compiler make them inconvenient to use in many cases (classObject and arrayObject in particular). I won't go into a full explanation here as this is not the right place for such a discussion. My intention here is just to figure out whether the APIs exposed can express more general or higher-level concepts so that we can limit the spread of existing Java-isms.

Reply:
@Leonardo2718 thanks for the clarification. I do agree some of these abstractions, in other parts of the compiler, are Java-specific. I do not think, however, that the limitations of VP constraints for other languages should be an issue to be solved here (as I am very sure you agree). My personal opinion here is that if we are going to use VP constraints to represent what we know about these abstract values to avoid the issues OpenJ9 had with a separate constraint infrastructure, then the API here would necessarily be coloured by the VP constraint structure. Does this add some inertia to changing the design of VP constraint? yes. Should we stop people using the infrastructure in a useful way in the project? I'd argue no. While VP constraints have notions of objects, arrays and strings and we may want to broaden them this API is not so specific to VP as to be useless if you chose to subclass and replace with some other kind of constraint infrastructure. I fear the alternative would be for the entire AbsValue implementation to be left to the language which I think would be less helpful than this proposal...

@xiacijie  feel free to chime in.

Reply:
@andrewcraik 

>  I do not think, however, that the limitations of VP constraints for other languages should be an issue to be solved here (as I am very sure you agree).

Completely agree.

> Should we stop people using the infrastructure in a useful way in the project? I'd argue no.

Also agree. However, even though the APIs are going to be "coloured" by VP constraints, I'd argue the extent of this should be minimized when possible. It's not clear to me that this has been achieved in the current design and I haven't seen any obvious explanation for why (e.g. alternative designs wouldn't work or are blocked by something, redesign work is in the plan, etc.).

> While VP constraints have notions of objects, arrays and strings and we may want to broaden them this API is not so specific to VP as to be useless if you chose to subclass and replace with some other kind of constraint infrastructure.

It's also not clear to me how this is intended to work. What things have to be sub-classed (`Abs*` classes, VP constraints, VP itself)? What needs to be overridden? There are not virtual functions currently, so how are overrides supposed to be used (I also don't see any use of static-polymorphism either)? This goes back to my earlier question about flexibility and customizability.

Reply:
Ok I think that helped me understand your concern better - @xiacijie could you provide the design alternatives considered and the extension mechanism envisioned for the AbsValues to address the valid concern above? Certainly adding some virtuals would seem to be necessary to allow extension...

Reply:
@Leonardo2718 See the latest post I made for the usage of VPConstaint.